# Тестовое задание НИУ ВШЭ

Задача №2

Рассчитайте с помощью формул Excel или языка программирования Python/R следующие показатели:
1. Страна, в которую было совершено больше всего поездок за рубеж и количество этих поездок
2. Подразделение, средняя длительность поездок сотрудников которого, наибольшая
3. Топ 5 подразделений московского кампуса ВШЭ по количеству участий в академической мобильности

! При решении задач использовались уже очищенные и скорректированные данные, столбец "Наименование подразделения" также уже добавлен.


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
import numpy as np
import datetime as dt

In [3]:
# настройка полного отображения столбца
pd.set_option('display.max_colwidth', -1)

# объявление переменной, которая содержит файл xlsx
file = "/content/drive/MyDrive/datasets/Test_Data_HSE.xlsx"
xlsx = pd.ExcelFile(file)

# парсим рабочий лист в переменную
df = xlsx.parse(xlsx.sheet_names[0])
df = df.convert_dtypes()

<ipython-input-3-157fe66d0295>:2: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('display.max_colwidth', -1)


In [4]:
df.head(5)

Код работника Дата рождения Категория персонала Ученая степень  \
0  OStaM-0308   NaT            НПР                 кандидат наук   
1  OStaM-1601   NaT            НПР                 PhD             
2  OStaM-0410   NaT            НПР                 нет степени     
3  OStaM-0510   NaT            НПР                 доктор наук     
4  OStaM-0787   NaT            НПР                 нет степени     

  Кампус работника Трудовые отношения с НИУ ВШЭ  \
0  Москва           штатный сотрудник             
1  Москва           штатный сотрудник             
2  Москва           штатный сотрудник             
3  Москва           штатный сотрудник             
4  Москва           штатный сотрудник             

  Код подразделения работника ВШЭ  \
0  05.03.                           
1  02.01.35.                        
2  02.33.03.                        
3  05.15.                           
4  05.15.                           

                                                Наименование подразделения  \
0  Институт проблем ценообразования и регулирования естественных монополий   
1  научно-учебная лаборатория исследований рынка труда                       
2  институт экономики транспорта и транспортной политики                     
3  Институт статистических исследований и экономики знаний                   
4  Институт статистических исследований и экономики знаний                   

                   Должность  \
0  старший научный сотрудник   
1  Доцент                      
2  научный сотрудник           
3  заведующий отделом          
4  научный сотрудник           

                                                                                                                                                                                                                                      Наименование программы  \
0  Научный семинар Национального исследовательского университета «Высшая школа экономики» по теме «Новые инструменты научно-технической и инновационной политики» и семинар Национального исследовательского университета «Высшая школа экономики» по теме «   
1  Научный семинар, Университет Болоньи «Производственные объединения малых и средних предприятий в Болонье, Модене и Реджио-Эмилия»                                                                                                                           
2  Научный семинар Национального исследовательского университета «Высшая школа экономики» по теме «Новые инструменты научно-технической и инновационной политики» и семинар Национального исследовательского университета «Высшая школа экономики» по теме «   
3  Научный семинар Национального исследовательского университета «Высшая школа экономики» по теме «Новые инструменты научно-технической и инновационной политики» и семинар Национального исследовательского университета «Высшая школа экономики» по теме «   
4  Курс лекций Национального исследовательского университета «Высшая школа экономики» по теме «Управление креативностью и инновациями» с привлечением ведущего зарубежного специалиста Яна Кратцера                                                            

                                                   Тип программы  \
0  Участие в научных мероприятиях (конференция, семинар, другое)   
1  Участие в научных мероприятиях (конференция, семинар, другое)   
2  Участие в научных мероприятиях (конференция, семинар, другое)   
3  Участие в научных мероприятиях (конференция, семинар, другое)   
4  Участие в научных мероприятиях (конференция, семинар, другое)   

   Стоимость участия в программе  \
0  0.0                             
1  12507.798738                    
2  24680.0                         
3  26992.07793                     
4  3060.32                         

                                                                         Документ, подтверждающий участие в программе   \
0  нет подтверждающего документа                                                                                    

In [5]:
df.dtypes

Код работника                                    string        
Дата рождения                                    datetime64[ns]
Категория персонала                              string        
Ученая степень                                   string        
Кампус работника                                 string        
Трудовые отношения с НИУ ВШЭ                     string        
Код подразделения работника ВШЭ                  string        
Наименование подразделения                       string        
Должность                                        string        
Наименование программы                           string        
Тип программы                                    string        
Стоимость участия в программе                    Float64       
Документ, подтверждающий участие в программе     string        
Страна, где организована программа               string        
Наименование принимающей организации             string        
Тип организации                         

## Задача №2

### 1. Страна, в которую было совершено больше всего поездок за рубеж и количество этих поездок

In [6]:
# подсчет поездок в страны, с сортировкой по убыванию по частоте
df_1 = df.groupby('Страна, где организована программа ')['Код работника'].count().reset_index(name='Количество поездок').sort_values(['Количество поездок'], ascending=False)
df_1.head()

Страна, где организована программа   \
21  Россия                                                       
23  США                                                          
4   Германия                                                     
27  Соединенное Королевство Великобритании и Северной Ирландии   
10  Италия                                                       

    Количество поездок  
21  405                 
23  25                  
4   11                  
27  10                  
10  7

США – страна, в которую было совершено больше всего поездок (25) за рубеж. На 2 месте Германия (11), третья Великобритания – 10 поездок.

2. Подразделение с наибольшей средней длительностью поездок сотрудников

In [7]:
# форматирование дат, приведение к формату даты "ГГГГ-ММ-ДД"
df['Дата начала программы']  = pd.to_datetime(df['Дата начала программы'].astype(str), format='%Y%m%d %H:%M:%S.%f')
df['Дата окончания программы']  = pd.to_datetime(df['Дата окончания программы'].astype(str), format='%Y%m%d %H:%M:%S.%f')

In [8]:
# добавление в исходный датафрейм столбца с количеством дней поездки
df['Длительность поездки'] = df['Дата окончания программы'] - df['Дата начала программы']
# форматирование значений в float (также прибавляем единицу для корректности значения длительности поездки)
df['Длительность поездки'] = df['Длительность поездки'].dt.days + 1
df['Длительность поездки']

0      3.0 
1      15.0
2      1.0 
3      1.0 
4      2.0 
      ...  
520    2.0 
521    2.0 
522    3.0 
523    4.0 
524    3.0 
Name: Длительность поездки, Length: 525, dtype: float64

In [9]:
# подсчет суммарных значений длительности поездок по подразделениям
df_SumDays = df.groupby(['Код подразделения работника ВШЭ', 'Наименование подразделения'])['Длительность поездки'].sum().reset_index(name='Сумма дней')
df_SumDays.head()

Код подразделения работника ВШЭ  \
0  01.06.03.                        
1  01.68.                           
2  01.71.                           
3  02.01.                           
4  02.01.31.                        

                                                                     Наименование подразделения  \
0  отдел правового сопровождения научной деятельности и вопросов интеллектуальной собственности   
1  Управление инновационной деятельности                                                          
2  Дирекция по экспертно-аналитической работе                                                     
3  Факультет экономических наук                                                                   
4  научно-учебная лаборатория корпоративных финансов                                              

   Сумма дней  
0  3.0         
1  11.0        
2  2.0         
3  4.0         
4  3.0

In [10]:
# подсчет количества поездок по подразделениям
df_ColExc = df.groupby('Код подразделения работника ВШЭ')['Код подразделения работника ВШЭ'].count().reset_index(name='Количество поездок')
df_ColExc

Код подразделения работника ВШЭ  Количество поездок
0   01.06.03.                       2                 
1   01.68.                          4                 
2   01.71.                          1                 
3   02.01.                          1                 
4   02.01.31.                       1                 
..        ...                      ..                 
73  07.03.15.                       1                 
74  10.01.                          11                
75  10.02.                          27                
76  10.03.                          14                
77  не указано                      39                

[78 rows x 2 columns]

In [11]:
# соединение таблиц в одну
df_2 = df_SumDays.merge(df_ColExc, on='Код подразделения работника ВШЭ', how='left')
df_2

Код подразделения работника ВШЭ  \
0   01.06.03.                        
1   01.68.                           
2   01.71.                           
3   02.01.                           
4   02.01.31.                        
..        ...                        
72  07.03.12.                        
73  07.03.15.                        
74  10.01.                           
75  10.02.                           
76  10.03.                           

                                                                      Наименование подразделения  \
0   отдел правового сопровождения научной деятельности и вопросов интеллектуальной собственности   
1   Управление инновационной деятельности                                                          
2   Дирекция по экспертно-аналитической работе                                                     
3   Факультет экономических наук                                                                   
4   научно-учебная лаборатория корпоративных финансов                                              
..                                                ...                                              
72  Институт налогового менеджмента и экономики недвижимости                                       
73  Высшая школа управления проектами                                                              
74  НИУ ВШЭ - Нижний Новгород                                                                      
75  НИУ ВШЭ - Санкт-Петербург                                                                      
76  НИУ ВШЭ - Пермь                                                                                

    Сумма дней  Количество поездок  
0   3.0         2                   
1   11.0        4                   
2   2.0         1                   
3   4.0         1                   
4   3.0         1                   
..  ...        ..                   
72  6.0         5                   
73  1.0         1                   
74  73.0        11                  
75  100.0       27                  
76  54.0        14                  

[77 rows x 4 columns]

In [12]:
# подсчет средней продолжительности поездки по подразделениям
df_2['Средняя продолжительность'] = df_2['Сумма дней'] / df_2['Количество поездок']
# сортировка по средней продолжительности по убыванию
df_2.sort_values('Средняя продолжительность', ascending=False, inplace=True)
df_2.head()

Код подразделения работника ВШЭ  \
12  02.04.22.                        
40  05.16.                           
7   02.01.35.                        
23  02.27.02.06.                     
6   02.01.34.                        

                                                                       Наименование подразделения  \
12  международная лаборатория по праву информационных технологий и интеллектуальной собственности   
40  Институт гуманитарных историко-теоретических исследований  им. А.В. Полетаева                   
7   научно-учебная лаборатория исследований рынка труда                                             
23  научно-учебная лаборатория психологии способностей                                              
6   научно-учебная лаборатория по финансовой инженерии и риск-менеджменту                           

    Сумма дней  Количество поездок  Средняя продолжительность  
12  35.0        1                   35.000000                  
40  50.0        3                   16.666667                  
7   15.0        1                   15.000000                  
23  14.0        1                   14.000000                  
6   9.0         1                   9.000000

Наибольшая средняя длительность поездок (35 дней) у Международной лаборатории по праву информационных технологий и интеллектуальной собственности.
Ожидаемо, подразделения с большой средней длительностью поездок имеют малое количество поездок – от 1 до 3.

3. Топ 5 подразделений московского кампуса ВШЭ по количеству участий в академической мобильности

In [47]:
# выявление подразделений с наибольшим количеством участий в академической мобильности
df_3 = df.groupby(['Код подразделения работника ВШЭ', 'Наименование подразделения'])['Код подразделения работника ВШЭ'].count().reset_index(name='Количество поездок').sort_values(['Количество поездок'], ascending=False)
df_3(6)

Код подразделения работника ВШЭ  \
35  05.03.                           
39  05.15.                           
41  05.19.                           
75  10.02.                           
44  05.28.                           
..     ...                           
20  02.26.12.                        
19  02.26.02.                        
18  02.26.                           
53  05.47.                           
56  05.52.                           

                                                    Наименование подразделения  \
35  Институт проблем ценообразования и регулирования естественных монополий      
39  Институт статистических исследований и экономики знаний                      
41  Институт государственного и муниципального управления                        
75  НИУ ВШЭ - Санкт-Петербург                                                    
44  Центр фундаментальных исследований                                           
..                                 ...                                           
20  научно-учебная лаборатория процессно-ориентированных информационных систем   
19  департамент анализа данных и искусственного интеллекта                       
18  Факультет компьютерных наук                                                  
53  Институт экономики природопользования и экологической политики               
56  Научно-учебная лаборатория исследований предпринимательства                  

    Количество поездок  
35  92                  
39  71                  
41  42                  
75  27                  
44  20                  
..  ..                  
20  1                   
19  1                   
18  1                   
53  1                   
56  1                   

[77 rows x 3 columns]

На топ-5 подразделений приходится 45,9% вовлечения в академическую мобильность - 241 участие из 525.
